openssl/thread_setup.c

```diff
@@ -0,0 +1,43 @@
+//go:build linux && !android
+// +build linux,!android
+
+#include "goopenssl.h"
+
+#include <stdio.h>
+#include <pthread.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+
+#define _GNU_SOURCE
+#include <unistd.h>
+ 
+#define CRYPTO_LOCK 0x01
+
+/* This array will store all of the mutexes available to OpenSSL. */ 
+static pthread_mutex_t *mutex_buf = NULL;
+ 
+static void locking_function(int mode, int n, const char *file, int line)
+{
+    if(mode & CRYPTO_LOCK)
+        pthread_mutex_lock(&mutex_buf[n]);
+    else
+        pthread_mutex_unlock(&mutex_buf[n]);
+}
+ 
+static unsigned long id_function(void)
+{
+    return (unsigned long)syscall(__NR_gettid);
+}
+ 
+int go_openssl_thread_setup(void)
+{
+    mutex_buf = malloc(go_openssl_CRYPTO_num_locks()*sizeof(pthread_mutex_t));
+    if(!mutex_buf)
+        return 0;
+    int i;
+    for(i = 0;  i < go_openssl_CRYPTO_num_locks(); i++)
```

Comment:
I didn't really mention that I also changed some spacing here to match style, so I think that bit got missed. (Adding a suggestion for each one I saw.)

```suggestion
    for (i = 0; i < go_openssl_CRYPTO_num_locks(); i++)
```